### DISCUSSION TITLE
Using Docker in your workspace

### DISCUSSION THREAD

**Discussion by janus-reith** [eclipse-theia/theia#7821] [General]
Most of my current projects involve multiple docker containers communicating with each other,
and I would preferrably interact with them using the integrated terminal.

How would I approach this best?
I could probably find some way to use the host docker instance, as e.g. portainer also uses the hosts docker from inside a container I guess this should work out some way.

However, as I would consider this a pretty typical usecase, I wonder if there is some straight-forward way to do this.

**Comment by akosyakov** [comment]
Hi, what do you mean by workspace? Theia is just an IDE. In this context, workspace means opened folder or multiple folders (multi-root). Do you use it in some product?

**Reply by janus-reith** [reply]
Sorry if that term is misleading. By workspace I mean the whole container environment that is available to me when starting a theia container, which only includes things specific to that container and whatever volumes I mapped from the host system.

I want to open theia, and then use the integrated terminal to start/stop/compose up docker containers.

Currently I'm using VS Code with the Remote SSH extension, so when I connect over ssh to a remote dev environment, the terminal is running in the scope of that host, where I can also start docker containers/do whatever I want in there.

Now if I start a container with theia on that remote dev host and use the integrated terminal, I only have access to what is available in that container.

**Reply by akosyakov** [reply]
Should not it be the same even if you don't run Theia in container, i.e. shell should be configured with docker cli and DOCKER_HOST env variable? Theia uses your env and shell configurations to spawn integrated terminals so it would provide same tooling.

Sorry I did not have exp with such setup, maybe @marcdumais-work can help?

**Reply by marcdumais-work** [reply]
I am also not very experienced with such setups - ATM we have single-container Theia applications, so this type of problems have not yet come-up for us in "base" Theia. 

I believe that Che has customizations,  that permit things similar to what @janus-reith is looking-for.  Maybe here? :https://github.com/eclipse/che-theia/blob/master/extensions/eclipse-che-theia-plugin-remote/src/node/terminal-container-aware.ts#L17

> However, as I would consider this a pretty typical usecase, I wonder if there is some straight-forward way to do this.

A bit more info:
By default Theia will use whatever is in `SHELL` environment variable when it comes time to spawn a terminal. To override, one can use `THEIA_SHELL` environment variable. See here: https://github.com/eclipse-theia/theia/blob/master/packages/terminal/src/node/shell-process.ts#L93
